### ISSUE TITLE
[Question] Have you ever consider leverage CSI or external-storage

### ISSUE THREAD

**Issue by gyliu513** [IBM/ubiquity#9]
Hi, there are two projects which may help this.

1) Kubernetes incubator project: https://github.com/kubernetes-incubator/external-storage
2) CSI: https://docs.google.com/document/d/1JMNVNP-ZHz8cGlnqckOnpJmHF-DNY7IYP-Di7iuVhQI/edit

Does anyone investigate those two projects?

**Comment by midoblgsm** [comment]
@gyliu513: https://github.com/IBM/ubiquity-k8s has an implementation of the external-storage.
For the CSI, we are following the progress and trying to contribute to it, since it is just a wip for the moment. Once CSI is agreed upon we will certainly implement it. For the moment we are using docker api (for DDC and Cloudfoundry) and Flex volume for k8s

**Comment by gyliu513** [comment]
Thanks @midoblgsm , another question is as Kubernetes already incubating [external storage]( https://github.com/kubernetes-incubator/external-storage), why not just leverage it for IBM ubiquity?

**Comment by midoblgsm** [comment]
@gyliu513, IBM ubiquity already leverages the external storage, we just wanted to separate the ubiquity service from the different frameworks implementations. We have the docker plugin implementation (https://github.com/IBM/ubiquity-docker-plugin) as well as the kubernetes implementation (https://github.com/IBM/ubiquity-k8s) that leverages the external storage as well as the flex volume  API for (attach/mount). Another implementation is for the CloudFoundry service broker (not yet checked in the public repo)

**Comment by gyliu513** [comment]
I'm interested in the CloudFoundry service broker, does it mean that this can work with [service catalog](https://github.com/kubernetes-incubator/service-catalog) to integrate with IBM BlueMix? @midoblgsm

**Comment by heikoludwig** [comment]
From a code perspective, yes. It could be integrated with any CloudFoundry service (using a recent Diego-based release), including Bluemix, by the operator of the service.

**Comment by gyliu513** [comment]
Great, I'm proposing a project here https://github.com/kubernetes-incubator/service-catalog/issues/548 and want to integrate BlueMix and Service Catalog, it would be great if I can check your Cloud Service Broker to get some help!

**Comment by jamescassell** [comment]
Now that CSI is a standard, does this project support it?  Can I plug this in place of e.g., GlusterFS?

**Comment by skrikler** [comment]
The Ubiquity project supports Kubernetes' Flex Volume. Yes, it can be used to plugin storage systems supported by Ubiquity to your Kubernetes cluster. Currently Ubiquity supports IBM storage systems - both block storage and file storage.
CSI will be supported in the future.

**Comment by jamescassell** [comment]
> The Ubiquity project supports Kubernetes' Flex Volume. Yes, it can be used to plugin storage systems supported by Ubiquity to your Kubernetes cluster. Currently Ubiquity supports IBM storage systems - both block storage and file storage.
> CSI will be supported in the future.

Thanks for the info!  Here's what I dug up:

From https://kubernetes.io/docs/concepts/storage/volumes/#flexVolume

> Flexvolume is an out-of-tree plugin interface that has existed in Kubernetes since version 1.2 (before CSI). It uses an exec-based model to interface with drivers. Flexvolume driver binaries must be installed in a pre-defined volume plugin path on each node (and in some cases master).
>
> Pods interact with Flexvolume drivers through the flexvolume in-tree plugin.

From https://github.com/kubernetes/kubernetes/issues/64984
> TLDR: CSI is positioned as our replacement for FlexVolumes.

CSI GA Announcement: https://kubernetes.io/blog/2019/01/15/container-storage-interface-ga/


Any idea when CSI will be supported?

Can I use this project as-is today to integrate a brand new k8s deployment with my existing SpectrumScale storage appliance?  Looking at either latest 1.14 vanilla k8s, or 3.11 OpenShift (latest; hopefully 4.0 if it comes out shortly).

**Comment by 27149chen** [comment]
> > The Ubiquity project supports Kubernetes' Flex Volume. Yes, it can be used to plugin storage systems supported by Ubiquity to your Kubernetes cluster. Currently Ubiquity supports IBM storage systems - both block storage and file storage.
> > CSI will be supported in the future.
> 
> Thanks for the info! Here's what I dug up:
> 
> From https://kubernetes.io/docs/concepts/storage/volumes/#flexVolume
> 
> > Flexvolume is an out-of-tree plugin interface that has existed in Kubernetes since version 1.2 (before CSI). It uses an exec-based model to interface with drivers. Flexvolume driver binaries must be installed in a pre-defined volume plugin path on each node (and in some cases master).
> > Pods interact with Flexvolume drivers through the flexvolume in-tree plugin.
> 
> From [kubernetes/kubernetes#64984](https://github.com/kubernetes/kubernetes/issues/64984)
> 
> > TLDR: CSI is positioned as our replacement for FlexVolumes.
> 
> CSI GA Announcement: https://kubernetes.io/blog/2019/01/15/container-storage-interface-ga/
> 
> Any idea when CSI will be supported?
> 
> Can I use this project as-is today to integrate a brand new k8s deployment with my existing SpectrumScale storage appliance? Looking at either latest 1.14 vanilla k8s, or 3.11 OpenShift (latest; hopefully 4.0 if it comes out shortly).

@jamescassell , sorry for the late update, CSI is ready: https://github.com/IBM/ibm-block-csi-operator
